pkg/config/configobserver/config_aggregator.go

```diff
@@ -0,0 +1,194 @@
+package configobserver
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/openshift/insights-operator/pkg/config"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+)
+
+type Interface interface {
+	Config() *config.InsightsConfiguration
+	ConfigChanged() (<-chan struct{}, func())
+	Listen(ctx context.Context)
+}
+
+// ConfigAggregator is an auxiliary structure that should obviate the need for the use of
+// legacy secret configurator and the new config map informer
+type ConfigAggregator struct {
+	lock               sync.Mutex
+	legacyConfigurator Configurator
+	configMapInformer  ConfigMapInformer
+	config             *config.InsightsConfiguration
+	listeners          map[chan struct{}]struct{}
+	usingInformer      bool
+	kubeClient         kubernetes.Interface
+}
+
+func NewConfigAggregator(ctrl Configurator, configMapInf ConfigMapInformer) Interface {
+	confAggreg := &ConfigAggregator{
+		legacyConfigurator: ctrl,
+		configMapInformer:  configMapInf,
+		listeners:          make(map[chan struct{}]struct{}),
+		usingInformer:      true,
+	}
+	confAggreg.mergeUsingInformer()
+	return confAggreg
+}
+
+// NewStaticConfigAggregator is a constructor used mainly for the techpreview configuration reading.
+// There is no reason to create and start any informer in the techpreview when data gathering runs as a job.
+// It is sufficient to read the config once when the job is created and/or starting.
+func NewStaticConfigAggregator(ctrl Configurator, cli kubernetes.Interface) Interface {
+	confAggreg := &ConfigAggregator{
+		legacyConfigurator: ctrl,
+		configMapInformer:  nil,
+		kubeClient:         cli,
+	}
+
+	confAggreg.mergeStatically()
+	return confAggreg
+}
+
+// mergeUsingInformer merges config values for the legacy "support" secret configuration and
+// from the new configmap informer. The "insights-config" configmap always takes
+// precedence if it exists and is not empty.
+func (c *ConfigAggregator) mergeUsingInformer() {
+	newConf := c.configMapInformer.Config()
+	conf := c.legacyConfigToInsightsConfiguration()
+
+	if newConf == nil {
+		c.config = conf
+		return
+	}
+
+	c.merge(conf, newConf)
+}
+
+func (c *ConfigAggregator) mergeStatically() {
+	conf := c.legacyConfigToInsightsConfiguration()
+	c.config = conf
+	cm, err := c.kubeClient.CoreV1().ConfigMaps("openshift-insights").Get(context.Background(), insightsConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return
+		}
+		klog.Error(err)
+	}
+
+	cmConf, err := readConfigAndDecode(cm)
+	if err != nil {
+		klog.Error("Failed to read configmap configuration: %v", err)
+		return
+	}
+
+	c.merge(conf, cmConf)
+}
+
+func (c *ConfigAggregator) merge(defaultCfg, newCfg *config.InsightsConfiguration) {
+	// read config map values and merge
+	if newCfg.DataReporting.Interval != 0*time.Minute {
```

Comment:
You can create a helper function here.

```go
func updateFieldIfNotEmpty(field interface{}, newValue interface{}) {
	fieldValue := reflect.ValueOf(field)
	newValueIsValid := !reflect.DeepEqual(newValue, reflect.Zero(reflect.TypeOf(newValue)).Interface())

	if newValueIsValid {
		fieldValue.Elem().Set(reflect.ValueOf(newValue))
	}
}
```
And replace all the Ifs with:

```go
updateFieldIfNotEmpty(&defaultCfg.DataReporting.Interval, newCfg.DataReporting.Interval)
updateFieldIfNotEmpty(&defaultCfg.DataReporting.UploadEndpoint, newCfg.DataReporting.UploadEndpoint)
// more...
```

Reply:
Yes it's an interesting suggestion, but I wouldn't complicate it with using the `reflect` package for now.

Reply:
Well, you could probably just create two separate functions:

```go
func updateDurationFieldIfNotEmpty(field *time.Duration, newValue time.Duration) {
	if newValue != 0*time.Minute {
		*field = newValue
	}
}

func updateStringFieldIfNotEmpty(field *string, newValue string) {
	if newValue != "" {
		*field = newValue
	}
}
```

Also, I'm pretty sure that you could compare the `Interval` to `0` without having to do `0*time.Minute`.

Reply:
Good point with zero, but I am not sure about the function suggestion and pointers usage. I think I still find it more straightforward as it is.....but yeah I am still thinking of a nicer solution.